// @sweetalert2/ngx-sweetalert2/esm5/swal-partial-targets.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import Swal from 'sweetalert2';
/**
 * Represents an object of targets for <swal> partials (use with *swalPartial directive).
 * We must use thunks to access the Swal.* functions listed below, because they get created after the first modal is
 * shown, so this object lets us reference those functions safely and in a statically-typed manner.
 */
var /**
 * Represents an object of targets for <swal> partials (use with *swalPartial directive).
 * We must use thunks to access the Swal.* functions listed below, because they get created after the first modal is
 * shown, so this object lets us reference those functions safely and in a statically-typed manner.
 */
SwalPartialTargets = /** @class */ (function () {
    function SwalPartialTargets() {
        this.title = function () { return Swal.getTitle(); };
        this.content = function () { return Swal.getContent(); };
        this.actions = function () { return Swal.getActions(); };
        this.confirmButton = function () { return Swal.getConfirmButton(); };
        this.cancelButton = function () { return Swal.getCancelButton(); };
        this.footer = function () { return Swal.getFooter(); };
    }
    return SwalPartialTargets;
}());
/**
 * Represents an object of targets for <swal> partials (use with *swalPartial directive).
 * We must use thunks to access the Swal.* functions listed below, because they get created after the first modal is
 * shown, so this object lets us reference those functions safely and in a statically-typed manner.
 */
export { SwalPartialTargets };
if (false) {
    /** @type {?} */
    SwalPartialTargets.prototype.title;
    /** @type {?} */
    SwalPartialTargets.prototype.content;
    /** @type {?} */
    SwalPartialTargets.prototype.actions;
    /** @type {?} */
    SwalPartialTargets.prototype.confirmButton;
    /** @type {?} */
    SwalPartialTargets.prototype.cancelButton;
    /** @type {?} */
    SwalPartialTargets.prototype.footer;
}
