// @sweetalert2/ngx-sweetalert2/esm5/sweetalert2.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { swalDefaultsProvider } from './di';
import { SwalPartialTargets } from './swal-partial-targets';
import { SwalPartialComponent } from './swal-partial.component';
import { SwalPartialDirective } from './swal-partial.directive';
import { SwalComponent } from './swal.component';
import { SwalDirective } from './swal.directive';
var SweetAlert2Module = /** @class */ (function () {
    function SweetAlert2Module() {
    }
    /**
     * @param {?=} defaultSwalOptions
     * @return {?}
     */
    SweetAlert2Module.forRoot = /**
     * @param {?=} defaultSwalOptions
     * @return {?}
     */
    function (defaultSwalOptions) {
        return {
            ngModule: SweetAlert2Module,
            providers: [swalDefaultsProvider(defaultSwalOptions)]
        };
    };
    SweetAlert2Module.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        SwalComponent, SwalPartialDirective, SwalPartialComponent,
                        SwalDirective
                    ],
                    providers: [
                        SwalPartialTargets
                    ],
                    imports: [
                        CommonModule
                    ],
                    exports: [
                        SwalComponent, SwalPartialDirective,
                        SwalDirective
                    ],
                    entryComponents: [
                        SwalComponent, SwalPartialComponent
                    ]
                },] }
    ];
    return SweetAlert2Module;
}());
export { SweetAlert2Module };
