// @sweetalert2/ngx-sweetalert2/esm2015/swal-partial-targets.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import Swal from 'sweetalert2';
/**
 * Represents an object of targets for <swal> partials (use with *swalPartial directive).
 * We must use thunks to access the Swal.* functions listed below, because they get created after the first modal is
 * shown, so this object lets us reference those functions safely and in a statically-typed manner.
 */
export class SwalPartialTargets {
    constructor() {
        this.title = () => Swal.getTitle();
        this.content = () => Swal.getContent();
        this.actions = () => Swal.getActions();
        this.confirmButton = () => Swal.getConfirmButton();
        this.cancelButton = () => Swal.getCancelButton();
        this.footer = () => Swal.getFooter();
    }
}
if (false) {
    /** @type {?} */
    SwalPartialTargets.prototype.title;
    /** @type {?} */
    SwalPartialTargets.prototype.content;
    /** @type {?} */
    SwalPartialTargets.prototype.actions;
    /** @type {?} */
    SwalPartialTargets.prototype.confirmButton;
    /** @type {?} */
    SwalPartialTargets.prototype.cancelButton;
    /** @type {?} */
    SwalPartialTargets.prototype.footer;
}
